Borders: Add BorderBoxControl component

aaronrobertshaw opened the pull request:
Related: 
- https://github.com/WordPress/gutenberg/issues/35602
- https://github.com/WordPress/gutenberg/pull/37769

## Description

This PR adds a `BorderBoxControl` component that provides a better experience when configuring individual borders for different sides.

- `BorderBoxControl`
    - This leverages the `BorderControl` to allow the configuration of a single flat border or individual borders for each side.
    - Can toggle between flat or split border views
    - Mixed values for different side borders are supported
    - Editing a single border property while others are mixed is supported e.g. you can set a consistent width from the linked or "flat" border view after having previously selected different colors and styles for each side. Each side border's width would be updated but their colors and styles would remain intact.

## How has this been tested?
Via unit tests;

1. Run`npm run test-unit packages/components/src/border-box-control/test/index.js` 
2. Run `npm run test-unit packages/components/src/border-box-control/test/utils.js`

Manually:

1. Checkout this PR branch
2. Run `npm run storybook:dev`
3. [Check out the `BorderBoxControl`](http://localhost:50240/?path=/story/components-experimental-borderboxcontrol--default)

## Screenshots

<img src="https://user-images.githubusercontent.com/60436221/151919488-4eeac561-e8c1-4a9f-b767-3cdcaac7a6fd.gif" width="500" />

## Types of changes
New feature.

## Known Issues

- Tooltips within the popover both for the close button and color palette swatches impact layout. Will be addressed in follow-up.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

aaronrobertshaw committed:
Add BorderBoxControl component

This provides a component through which you can configure separate borders for individual sides.

mirka commented on packages/components/src/border-box-control/border-box-control-split-controls/component.tsx:
```diff
@@ -0,0 +1,77 @@
+				onChange={ ( newBorder ) => onChange( newBorder, 'top' ) }
+				value={ value?.top }
+				{ ...sharedBorderControlProps }
+			/>
```
I think we need some accessible labeling on each of these so people can tell which side the border is for.

aaronrobertshaw replied:
I've added visually hidden labels for each of these split controls in https://github.com/WordPress/gutenberg/pull/38876/commits/310cb965cf4ae28cd246c57999617143fb805d2a. 

Do you think this is sufficient for a first pass on this component?

mirka replied:
Absolutely!

mirka commented on packages/components/src/border-box-control/styles.ts:
```diff
@@ -0,0 +1,44 @@
+	position: relative;
+	gap: ${ space( 4 ) };
+	flex: 1;
+	margin-right: ${ space( 3 ) };
```
I believe this is responsible for the missing margin here?

<img src="https://user-images.githubusercontent.com/555336/155979460-ce510891-1b8c-49d2-848c-c93ff0d0a2af.png" alt="Missing margin in RTL" width="200">

aaronrobertshaw replied:
I've added RTL styles via https://github.com/WordPress/gutenberg/pull/38876/commits/fb6833ea0c16388d8e212b2f1d6724872a52cda5

mirka commented on packages/components/src/border-box-control/border-box-control-visualizer/component.tsx:
```diff
@@ -0,0 +1,35 @@
+		borderRight: getClampedWidthBorderStyle( value?.right ),
+		borderBottom: getClampedWidthBorderStyle( value?.bottom ),
+		borderLeft: getClampedWidthBorderStyle( value?.left ),
+	};
```
In RTL there is a mismatch between the visualizer and the input placement:

<img src="https://user-images.githubusercontent.com/555336/155979852-17fd31e8-2687-499b-b1ab-80ebaa0af7b4.png" alt="In unlinked mode in RTL, the left edge visualizer changes when the right edge input value is changed" width="200">

aaronrobertshaw replied:
Thanks! I've fixed this up while also moving the inline styles to a dynamic class.

mirka commented on packages/components/src/border-box-control/utils.ts:
```diff
@@ -0,0 +1,158 @@
+	};
+};
+
+export const getShorthandBorderStyle = ( border?: Border ) => {
```
I'm not quite sure if this was intentional — using shorthand for partial values potentially causes the omitted value to actually be overridden. For example in Chrome here, the width is reverted to `initial`, which is different from the originally set 1px. I can imagine there would also be cross-browser differences around this.

<img src="https://user-images.githubusercontent.com/555336/155986639-543ac7e1-8dad-48dc-9a4b-4740757024e4.png" alt="Omitting the width value from the border shorthand causes it to become initial" width="350">

aaronrobertshaw replied:
Good catch 👍 

I've added an optional fallback border that can be used to flesh out omitted values however the consumer wants. 

The current uses of the util didn't appear to be impacted by the previous oversight.

mirka commented on packages/components/src/border-box-control/styles.ts:
```diff
@@ -0,0 +1,44 @@
+`;
+
+export const BorderBoxControlVisualizer = css`
+	border: ${ CONFIG.borderWidth } solid ${ COLORS.gray[ 200 ] };
```
I was initially a bit confused by how the visualizer behaves in a not-quite-WYSIWYG way, but thinking through it, I understood that it was because we need to support "partial" values, like `width` only. It's pretty clever, once I got it!

<img width="288" alt="Border width zero makes the visualizer border disappear" src="https://user-images.githubusercontent.com/555336/155982425-cd9a99b1-d605-46d5-9f03-5e3ce0eb25c7.png">
<img width="292" alt="Changing the border color changes the visualizer border color" src="https://user-images.githubusercontent.com/555336/155982506-e2c700c5-dac7-42d5-ba0b-36f5538881d7.png">
<img width="295" alt="Just setting the border width changes the visualizer border color to black (from a light gray)" src="https://user-images.githubusercontent.com/555336/155982586-23033053-67c0-4b51-9232-170d68c15c33.png">

---

Is there a reason I'm not allowed to set a `style` only?

<img src="https://user-images.githubusercontent.com/555336/155985556-27efe049-60d3-4767-87bc-45545e627988.gif" alt="CleanShot 2022-02-28 at 21 43 27" width="200">

aaronrobertshaw replied:
> I was initially a bit confused by how the visualizer behaves in a not-quite-WYSIWYG way, but thinking through it, I understood that it was because we need to support "partial" values, like `width` only. It's pretty clever, once I got it!

I've tried my best to make it intuitive. Reflecting a state where no selections have been made and inheritance is possible is a little tricky. 

As your screenshots show, when explicitly setting a zero width border we can reflect the fact there is no border. With only a color selected we do have a border just only with inherited width and style. Similarly, setting only the width to a non-zero number we need to handle an inherited color and style.

I'm definitely open to improvements. For now I'd be leaning to any improvements being handled via a separate PR.


> Is there a reason I'm not allowed to set a `style` only?

This came from other discussions around the border supports where border style was determined to be "secondary" to width and color. Part of that was that the style should be cleared whenever there is no border color or width. This in turn effectively means you can't set a border style only.

The `BorderControl` component itself has a prop that allows this sanitisation to be turned off. 

If it proves desirable, we could provide this on the `BorderBoxControl` as well in a follow-up.

mirka replied:
> I'm definitely open to improvements.

I think it's good! There are no logical inconsistencies, at the very least.

> This in turn effectively means you can't set a border style only.

I see. I'll just note that:

- In unlinked mode you can't set a border style only, but in linked mode you can.
- If a border style cannot be set for whatever reason, should that be communicated accessibly instead of silently ignoring the clicks? (Like by disabling the buttons?)

Feel free to move to a follow-up 🙂

aaronrobertshaw replied:
From memory, the reason you could set a style only on the linked control was that we might be dealing with mixed borders. For example, if the individual sides have different colors, styles and widths, then the linked control would have undefined values for those. Setting only a style here would then apply that to the individual sides.

https://user-images.githubusercontent.com/60436221/159430835-1dde7393-35ec-4f98-9a9a-34b12853f5be.mp4

I agree that we can refine this in a follow-up.

mirka reviewed (commented):
This is a very cool component! While there is a good amount of complexity being handled on the logic side, I was pleasantly surprised at how clean the UI implementation itself was. I don't know why but before reading the code I totally overcomplicated the visualizer implementation in my head — but it's actually just a positioned div! 🤦😆 Elegant use of `grid` on top of it too!

ciampo commented on packages/components/src/border-box-control/border-box-control-visualizer/component.tsx:
```diff
@@ -0,0 +1,35 @@
+	forwardedRef: React.Ref< any >
+) => {
+	const { value, ...otherProps } = useBorderBoxControlVisualizer( props );
+	const styles = {
+		borderTop: getClampedWidthBorderStyle( value?.top ),
+		borderRight: getClampedWidthBorderStyle( value?.right ),
+		borderBottom: getClampedWidthBorderStyle( value?.bottom ),
+		borderLeft: getClampedWidthBorderStyle( value?.left ),
+	};
+
+	return <View { ...otherProps } ref={ forwardedRef } style={ styles } />;
```
Can we avoid using inline styles here? We could go for an approach similar to what discussed in https://github.com/WordPress/gutenberg/pull/37769#discussion_r809362205

aaronrobertshaw replied:
Thanks for the suggestion 🙇 

I've updated this to apply the visualizer's border styles via a dynamic class.

ciampo reviewed (commented):
As others have also commented, this is great work!

I'm just leaving here some high-level feedback that I also gave on the sister PR #37769:

- using `ForwardedRef` type instead of `Ref`
- wrapping all styles that affect right/left in the `rtl()` utility function
- trying to replace as many hardcoded DOM structures as possible with the Styled component selector ([example](https://github.com/WordPress/gutenberg/pull/37769#discussion_r809380148))
- Investigate building the UI with the `Flex` and `Grid` components when possible over CSS styles (this is more of a preference / exploration, definitely not blocking — I think it's interesting to explore the composability of our components and discover any potential limitations!)
- avoid using inline styles where possible

Thank you again for your work on this one!

aaronrobertshaw committed:
Add RTL styles for split border controls

aaronrobertshaw committed:
Use ForwardRef for typing instead of Ref

aaronrobertshaw committed:
Move inline visualizer styles to dynamic class

aaronrobertshaw committed:
Add RTL styles for BorderBoxControlVisualizer

aaronrobertshaw committed:
Avoid using hardcoded DOM element

aaronrobertshaw committed:
Use Grid for split controls layout

aaronrobertshaw committed:
Prevent application of value prop on visualizer element

aaronrobertshaw committed:
Refactor clamped and shorthand border style generation

- Moves the `getClampedWidthBorderStyle` functionality into the style.ts file
- Updates `getShorthandBorderStyle` to accept optional fallback border styles to  work around possible override of inherited values.

aaronrobertshaw committed:
Update BorderBoxControl tests to match latest aria labels

aaronrobertshaw committed:
Add visually hidden labels for split border controls

mirka reviewed (approved):
Awesome work on this complex component! 🚀

aaronrobertshaw committed:
Add SlotFill provider as wrapper to storybook example
